x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java

```diff
@@ -279,6 +284,79 @@ private static void checkAggregate(LogicalPlan p, Set<Failure> failures) {
         }
     }
 
+    /**
+     * Check CATEGORIZE grouping function usages.
+     * <p>
+     *     Some of those checks are temporary, until the required syntax or engine changes are implemented.
+     * </p>
+     */
+    private static void checkCategorizeGrouping(Aggregate agg, Set<Failure> failures) {
+        // Forbid CATEGORIZE grouping function with other groupings
+        if (agg.groupings().size() > 1) {
+            agg.groupings().forEach(g -> {
+                g.forEachDown(
+                    Categorize.class,
+                    categorize -> failures.add(
+                        fail(categorize, "cannot use CATEGORIZE grouping function [{}] with multiple groupings", categorize.sourceText())
+                    )
+                );
+            });
+        }
+
+        // Forbid CATEGORIZE grouping function within other functions
+        agg.groupings().forEach(g -> {
+            g.forEachDown(
+                Function.class,
```

Comment:
This is not quite general enough. E.g. Binary operators also need to be forbidden, for instance, otherwise stuff like `BY CATEGORIZE(field) == "some string"` will be permitted.

The _positive_ pattern to look for is that `CATEGORIZE` can only be the direct child of the topmost `Alias`. If there is a `CATEGORIZE` but it's not the top-level expression in the `Alias`, there must be a function, operator etc. that contains it.

Reply:
After a quick check, operators are BinaryPredicate (sub_^N_-classes), so I think it's covered.

About the Alias pattern, I can change it. I wasn't sure if there was some kind of weird syntax I didn't know, like `MV_COUNT(x = CATEGORIZE(a))` (Just an example xD).

If you think there's no other alias notation that could appear, I can change it! It just looked "safer" this way

Reply:
You're right!

I noticed this doesn't provide the correct error message, but it turns out this is due to `CATEGORIZE` output data type being `Integer` at the moment:
```
curl -u elastic:password -H "Content-Type: application/json" "127.0.0.1:9200/_query?format=txt" -d '{                                                             
    "query": "row x = \"asd\" | stats values(x) by categorize(x) == \"fdsa\""
}'
{"error":{"root_cause":[{"type":"verification_exception","reason":"Found 1 problem\nline 1:36: first argument of [categorize(x) == \"fdsa\"] is [numeric] so second argument must also be [numeric] but was [keyword]"}],"type":"verification_exception","reason":"Found 1 problem\nline 1:36: first argument of [categorize(x) == \"fdsa\"] is [numeric] so second argument must also be [numeric] but was [keyword]"},"status":400}
```

Let's just add the test case for good measure?

I think looking for the Alias pattern is more likely to remain true as we really only use aliases in case of assignments and renames - resp. only for assignments in STATS. But both ways should work so you get to chose!